File: cli/gh-teacher/internal/configrepo/secret_test.go
package configrepo

import "testing"

func TestGenerateSecret_MatchesPatternAndLength(t *testing.T) {
	for _, n := range []int{4, 8, 16, 64} {
		s, err := GenerateSecret(n)
		if err != nil {
			t.Fatalf("GenerateSecret(%d) error: %v", n, err)
		}
		if len(s) != n {
			t.Errorf("GenerateSecret(%d) length = %d, want %d", n, len(s), n)
		}
		if !SecretPattern.MatchString(s) {
			t.Errorf("GenerateSecret(%d) = %q does not match SecretPattern", n, s)
		}
		if err := ValidateSecret(s); err != nil {
			t.Errorf("generated secret %q failed ValidateSecret: %v", s, err)
		}
	}
}

func TestGenerateSecret_RejectsNonPositiveLength(t *testing.T) {
	if _, err := GenerateSecret(0); err == nil {
		t.Error("GenerateSecret(0) = nil error, want error")
	}
	if _, err := GenerateSecret(-3); err == nil {
		t.Error("GenerateSecret(-3) = nil error, want error")
	}
}

func TestGenerateSecret_IsRandom(t *testing.T) {
	// Two draws of a reasonable length should differ with overwhelming
	// probability; a constant generator would be a real bug.
	a, _ := GenerateSecret(DefaultSecretLength)
	b, _ := GenerateSecret(DefaultSecretLength)
	if a == b {
		t.Errorf("two GenerateSecret(%d) draws were identical (%q) — not random", DefaultSecretLength, a)
	}
}

func TestValidateSecret(t *testing.T) {
	valid := []string{"abcd", "abc123", "0a0a0a0a", "zzzz9999"}
	for _, s := range valid {
		if err := ValidateSecret(s); err != nil {
			t.Errorf("ValidateSecret(%q) = %v, want nil", s, err)
		}
	}

	invalid := []string{
		"",        // empty
		"abc",     // too short (< 4)
		"ABC123",  // uppercase
		"abc-123", // hyphen (would split the path segment)
		"abc/123", // slash (path traversal vector)
		"abc 123", // space
		"abc.123", // dot
	}
	for _, s := range invalid {
		if err := ValidateSecret(s); err == nil {
			t.Errorf("ValidateSecret(%q) = nil, want error", s)
		}
	}
}

File: src/util/secret.test.ts
import { describe, expect, it } from "vitest"
import {
  DEFAULT_SECRET_LENGTH,
  SECRET_PATTERN,
  generateSecret,
  isValidSecret,
} from "./secret"

describe("generateSecret", () => {
  it("produces a value matching the pattern at the requested length", () => {
    for (const n of [4, 8, 16, 64]) {
      const s = generateSecret(n)
      expect(s).toHaveLength(n)
      expect(SECRET_PATTERN.test(s)).toBe(true)
      expect(isValidSecret(s)).toBe(true)
    }
  })

  it("defaults to DEFAULT_SECRET_LENGTH", () => {
    expect(generateSecret()).toHaveLength(DEFAULT_SECRET_LENGTH)
  })

  it("throws on a non-positive length", () => {
    expect(() => generateSecret(0)).toThrow()
    expect(() => generateSecret(-2)).toThrow()
  })

  it("is random across draws", () => {
    expect(generateSecret(12)).not.toBe(generateSecret(12))
  })
})

describe("isValidSecret", () => {
  it("accepts safe path segments", () => {
    for (const s of ["abcd", "abc123", "0a0a0a0a", "zzzz9999"]) {
      expect(isValidSecret(s)).toBe(true)
    }
  })

  it("rejects empty, too-short, uppercase, or separator-bearing values", () => {
    for (const s of [
      "",
      "abc",
      "ABC123",
      "abc-123",
      "abc/123",
      "abc 123",
      "abc.123",
    ]) {
      expect(isValidSecret(s)).toBe(false)
    }
  })
})
